master/pkg/schemas/defaults.go

```diff
@@ -0,0 +1,229 @@
+package schemas
+
+import (
+	"encoding/json"
+	"fmt"
+	"reflect"
+	"strings"
+)
+
+// RuntimeDefaultable means there are runtime values for filling in an object, like choosing a
+// random seed based on the wall clock.
+type RuntimeDefaultable interface {
+	// RuntimeDefaults must apply the runtime-defined default values.
+	RuntimeDefaults()
+}
+
+// Defaultable means a struct can have its defaults filled in automatically.
+type Defaultable interface {
+	// DefaultSource must return a parsed json-schema object in which to find defaults.
+	DefaultSource() interface{}
+}
+
+// FillDefaults will recurse through structs, maps, and slices, setting default values for any
+// struct fields whose struct implements the Defaultable interface.  This lets us read default
+// values out of json-schema automatically.
+//
+// There are some forms of defaults which must be filled at runtimes, such as giving a default
+// description to experiments with no description.  This can be accomplished by implementing
+// the RuntimeDefaultable interface for that object.  See ExperimentConfig for an example.
+//
+// There are some objects which get their defaults from other objects' defaults.  This an
+// unforutnate detail of our union types which have common members that appear on the root union
+// object.  That's hard to reason about, and we should avoid doing that in new config objects.  But
+// those objects implement DefaultSource() to customize that behavior.
+//
+// Example usage:
+//
+//    config, err := expconf.ParseAnyExperimentConfigYAML(bytes)
+//
+//    // Use the cluster checkpoint storage if the user did not specify one.
+//    schemas.Merge(&config.CheckpointStorage, cluster_default_checkpoint_storage)
+//
+//    // Define any remaining undefined values.
+//    schemas.FillDefaults(&config)
+//
+func FillDefaults(obj interface{}) {
+	vObj := reflect.ValueOf(obj)
+	// obj can't be a non-pointer, because it edits in-place.
+	if vObj.Kind() != reflect.Ptr {
+		panic("FillDefaults must be called on a pointer")
+	}
+	// obj can't be a nil pointer, because FillDefaults(nil) doesn't make any sense.
+	if vObj.IsZero() {
+		panic("FillDefaults must be called on a non-nil pointer")
+	}
+	// Enter the recursive default filling with no default bytes for the root object (which must
+	// already exist), and starting with the name of the object type.
+	name := fmt.Sprintf("%T", obj)
+	vObj.Elem().Set(fillDefaults(vObj.Elem(), nil, false, name))
+}
+
+func fillDefaults(obj reflect.Value, defaultBytes []byte, inPtr bool, name string) reflect.Value {
```

Comment:
Sure.

Reply:
Wait... no, because `fillDefaults` typically returns the original object, like in the case of a struct or pointer.

Reply:
I'd be ok rewriting `fillDefaults` to return a copy of the data if you wanted that behavior, and then renaming it.  I don't see that it much matters until the generics land, in which case we would expose a function like:

```go
func WithDefaults[T any](obj T) T {
   ...
}
```

Reply:
> Wait... no, because fillDefaults typically returns the original object, like in the case of a struct or pointer

Ah yeah.. it just looks like it doesn't. Lol.